JSDK-1510 Implementing the Bandwidth Constraints example app.

manjeshbhargav opened the pull request:
@markandrus 

This PR contains the implementation of the Bandwidth Constraints Example.

Screenshot:
![screen shot 2017-09-18 at 5 32 05 pm](https://user-images.githubusercontent.com/755797/30570627-370d98f6-9c97-11e7-9893-e28dee19f668.png)

markandrus commented on README.md:
```diff
@@ -45,8 +45,10 @@ video in both the tabs!
 
-The project contains some common use-case examples for the Twilio Video JS SDK.
+The project contains some common use-case examples for the Twilio Video JS SDK. After running the application
+by following the instructions above, click on these links to run the examples.
```
I think this indicates we ought to find a way to include these links in the app, e.g. a navigation bar or something?

manjeshbhargav replied:
@markandrus I'll add an `examples/index.html` page linking to all the examples.

markandrus commented on examples/bandwidthconstraints/src/index.js:
```diff
@@ -0,0 +1,146 @@
+  if (!canvas) {
+    waveformContainer.appendChild(waveform.element);
+  }
+}
```
We need to start factoring this stuff out, otherwise it will get difficult to maintain, IMO.

manjeshbhargav replied:
@markandrus I've already moved `Waveform` to the `utils` folder.

markandrus replied:
How about our "attach"/"detach" helpers? Can these be made common?

manjeshbhargav replied:
@markandrus They're not entirely common. In this case, we're also attaching the Track to the Waveform module, whereas in other examples we are not. And I also don't want too much of the app logic to be modularized as it affects the flow of reading the code.

markandrus commented on examples/bandwidthconstraints/src/index.js:
```diff
@@ -0,0 +1,146 @@
+  // display the media of the second Participant that will enter
+  // the Room with bandwidth constraints.
+  return Video.connect(creds.token, { tracks: [] });
+}).then(function(_room) {
```
I think we can use `async`/`await` now. This also lets us avoid binding `_room`.

manjeshbhargav replied:
@markandrus Do Firefox and Safari support it?

markandrus replied:
Yes, recent versions support it.

markandrus commented on examples/util/getroomcredentials.js:
```diff
@@ -0,0 +1,20 @@
+        resolve(JSON.parse(xhr.responseText));
+      }
+    };
+    xhr.send(null);
```
Can we use `fetch` instead?

manjeshbhargav replied:
@markandrus Do Firefox and Safari support it?

markandrus replied:
Yes

markandrus commented:
Looks good. What about indicating to the user that the constraints took effect? E.g., by reading `getStats`? Will it visually be clear enough that the video is at a lower bitrate? Audibly?

manjeshbhargav commented:
@markandrus I'm working on adding the bitrate graphs. I'm thinking of leveraging the graph module used by the webrtc sample app demonstrating bandwidth constraints: https://github.com/webrtc/samples/blob/gh-pages/src/js/third_party/graph.js
WDYT?

manjeshbhargav commented:
@markandrus I've updated the PR with the following changes:

1. Used `async/await` in the Bandwidth Constraints example app.
2. Used `fetch()` for getting the Room token.
3. Added `examples/index.html` that lists the example apps' links.
4. Added bitrate graphs in the Bandwidth Constraints example app.

Screenshot of `examples/index.html`:
![screen shot 2017-09-19 at 3 52 55 pm](https://user-images.githubusercontent.com/755797/30619531-c5c0bb60-9d54-11e7-9be7-4475f47a6ea3.png)

Screenshot of the bitrate filters:
![screen shot 2017-09-19 at 12 38 58 pm](https://user-images.githubusercontent.com/755797/30619553-de5af6c2-9d54-11e7-89ce-abca36d00767.png)

markandrus commented on examples/bandwidthconstraints/src/index.js:
```diff
@@ -0,0 +1,214 @@
+var helpers = require('./helpers');
+var waveform = require('../../util/waveform');
+var connectWithBandwidthConstraints = helpers.connectWithBandwidthConstraints;
+var updateBandwidthConstraints = helpers.updateBandwidthConstraints;
```
Can you also use `const` and `let` instead of `var`? If `async`/`await` is supported, so is `const` and `let`.

markandrus commented on examples/bandwidthconstraints/src/index.js:
```diff
@@ -0,0 +1,214 @@
+  // Connect to a random Room with no media. This Participant will
+  // display the media of the second Participant that will enter
+  // the Room with bandwidth constraints.
+  var _room = await Video.connect(creds.token, { tracks: [] });
```
Why are we binding `_room` instead of assigning to `room`?

markandrus replied:
Ah, why not `randomRoom` or `someRoom`?

markandrus commented on examples/bandwidthconstraints/src/index.js:
```diff
@@ -0,0 +1,214 @@
+async function connectToOrDisconnectFromRoom(e) {
+  e.preventDefault();
+  if (room) {
+    e.target.value = 'Connect to Room';
```
You already have `connectOrDisconnect`, can you set `connectOrDisconnect.value` instead of `e.target`?

markandrus commented on examples/bandwidthconstraints/src/index.js:
```diff
@@ -0,0 +1,214 @@
+/**
+ * Connect to or disconnect from a Room.
+ */
+async function connectToOrDisconnectFromRoom(e) {
```
I think `event` is better than `e` here.

markandrus commented on examples/bandwidthconstraints/src/index.js:
```diff
@@ -0,0 +1,214 @@
+}
+
+/**
+ * Connect to or disconnect from a Room.
```
I think it is better to have two functions, `connectToRoom`, and `disconnectFromRoom`, then let `connectToRoom` reassign the "click" event handler of  to `disconnectFromRoom` and vice-versa.

markandrus commented on examples/bandwidthconstraints/src/index.js:
```diff
@@ -0,0 +1,214 @@
+        ? stats[0].remoteAudioTrackStats[0]
+        : stats[0].remoteVideoTrackStats[0]
+      var _bytesReceived = remoteTrackStats.bytesReceived;
+      var _timestamp = remoteTrackStats.timestamp;
```
Why are we binding `_bytesReceiver` and `_timestamp`?

markandrus commented on examples/util/getroomcredentials.js:
```diff
@@ -0,0 +1,13 @@
+function getRoomCredentials() {
+  return fetch('/token').then(function(response) {
+    return response.json();
+  });
```
Can use fat-arrow functions and `async`/`await`. Basically, I would like us to use all the modern syntax since it is widely available now, e.g.

```
async function getRoomCredentials() {
  const response = await fetch('/token')
  return response.json();
}
```

markandrus commented on examples/util/timelinegraph.js:
```diff
@@ -0,0 +1,697 @@
+ *  Copyright (c) 2015 The WebRTC project authors. All Rights Reserved.
+ *
+ *  Use of this source code is governed by a BSD-style license
+ *  that can be found in the LICENSE file in the root of the source
```
We need to add/update our LICENSE.md file. It can take the same form as https://github.com/twilio/twilio-video.js/blob/master/LICENSE.md with a couple changes:

1. First section: Twilio license
2. Second section: the WebRTC license

markandrus commented:
One comment, otherwise looks good 👍
